tests/unit-tests/#.scm

```diff
@@ -65,14 +66,28 @@
           (##not (##number? n2))
           (##< tolerance (##magnitude (##- n1 n2))))
       (##failed-check msg n1)))
-                  
+
 (##define-macro (##setup-check)
 
   (eval '(##begin
 
            (define ##enable-checks?
              (make-parameter #t))
 
+           (define ##tested-procedures '())
+
+           (define (##proc? name)
+             (##procedure? (##global-var-ref (##make-global-var name))))
+
+           (define (##record-tested-procedures! expr)
```

Comment:
Can you explain what this does?  I assume that when there is a `(check-XXX (FOO ...) ...)` it will infer that `FOO` is a primitive that is being tested.  This is often true, but what about `(check-XXX (let ...) ...)` and `(check-XXX (helper) ...)` where `helper` is a procedure that does various operations.

Here is a real example: `04-list/carpluscdr.scm` contains this test:

```
(check-equal? (call-with-values (lambda () (car+cdr a0)) list) '(0 1))
```

So it doesn't know that it is `car+cdr` that is being tested.  It may be useful to handle a few common patterns such as `(let[*] ((<var1> <expr1>) (<var2> <expr2>) ...) <expr0>)` where all the `<exprN>` are searched for primitives that are called.  Things like `call-with-values`, `and`, `or`, and `begin` would need to be handled too.  If someone writes a test that does not fall into one of those cases then the call to the primitive would not be registered and the system would say that that there is no unit test coverage of that primitive (unless some other tests exist).  This would prompt the unit test writer to either modify how the test is done, or to extend the logic to detect that the test does call the primitive.

Reply:
It check for immediate call. It indeed won't find call that are nested in special form (ex. `let`, `and`, `or`).
I am not sure which call is really tested in some of the special form.

Can we consider that both `call-with-values` and `car+cdr` are tested in the test or it's just `car+cdr`:
```
(check-equal? (call-with-values (lambda () (car+cdr a0)) list) '(0 1))
```

Reply:
I found out how to get the source code of the helper. By using the `##decompile` function which returns the code of the procedure or `#f` otherwise.

To fetch the global variable I use `(##global-var-ref (##make-global-var sym))`, but it fails to find the procedure `str->obj` in the test `09-io/six.scm`. The value returned is `#!unbound`.
 
```
(define (str->obj str)
  (with-exception-catcher
   (lambda (exc)
     ;;(display-exception exc)
     exc)
   (lambda ()
     (with-input-from-string str read-all))))

(define-macro (chk str first . rest)
  `(check-equal? (str->obj ,str) '((six.infix ,first) ,@rest)))

(chk "\\a"        (six.identifier a))
```

Reply:
I don't know why `##global-var-ref` is failing to read `str->obj`.  What is `##decompile` returning?

By the way there is also `##global-var-primitive-ref` to fetch the standard binding of a global variable.  For example:

```
> (##global-var-ref (##make-global-var 'car))          
#<procedure #2 car>
> (set! car 42)                                        
> (##global-var-ref (##make-global-var 'car))
42
> (##global-var-primitive-ref (##make-global-var 'car))
#<procedure #2>
> ((##global-var-primitive-ref (##make-global-var 'car)) '(11 22 33))
11
```

Reply:
`##decompile` returns the list representing the source code of the helper function. This is use to fetch primitives inside helper function.
```
> (define (obj->str obj printer style)
  (with-output-to-string
   ""
   (lambda ()
     (output-port-readtable-set!
      (current-output-port)
      (readtable-sharing-allowed?-set
       (output-port-readtable (current-output-port))
       style))
     (printer obj))))
> (##decompile obj->str)
(lambda (obj printer style)
  (with-output-to-string
   ""
   (lambda () (output-port-readtable-set! (current-output-port) (readtable-sharing-allowed?-set (output-port-readtable (current-output-port)) style)) (printer obj))))
```

Does the eval in `##setup-check` allows the use of the definitions in macros?
```
(##define-macro (##setup-check)
  `(val` '(##begin …
```

The following macro extends the scope of the define to the macros' realm. This seems to solve the issue with `obj->str` not being found.
```
(##define-macro (define ident-or-pair #!rest body)
 (let ((code `(##define ,ident-or-pair ,@body)))
   (if (symbol? ident-or-pair)
     code
     (begin
       (eval code)
       '(##begin)))))
```

Now the remaining issue is a recursion.in `09-io/write_shared.scm`. The recording of tested procedure has to remember the helper functions that were already recorded. My code doesn't handle recursive helper function yet.

Reply:
From your macro definition of `define` I suspect that the reason you could not decompile `obj->str` is that you have a phasing problem (mixing the expansion time and the run time).  The call to `##decompile` needs to be done at run time, not at expansion time.

Reply:
Another approach would be to use the interpreter's single-stepping mechanism to keep track of all the procedures that are called.  Here is a prototype of how this can be done:

```
;;;----------------------------------------------------------------------------

;;; The following code will change the interpreter's call handler so
;;; that whenever the interpreter executes a procedure call it will
;;; register the called procedure in a table to keep track of the set
;;; of procedures that are called.

(define procedures-called (make-table))

(let ((stepper (##current-stepper)))
  (vector-set!
   stepper
   1 ;; index of call handler in stepper
   (parameterize ((##current-stepper (vector-copy stepper)))

     ;; need to create this call handler in the context of a new
     ;; stepper so that the calls it executes will not invoke the call handler

     (eval '(lambda (leapable? $code rte execute-body . other)
              (let ((proc (car other))) ;; procedure being called
                (table-set! procedures-called
                            proc
                            (+ 1 (table-ref procedures-called proc 0)))
                (apply execute-body (cons $code (cons rte other)))))))))

;;;----------------------------------------------------------------------------

;; code being tested:
(let loop ((i 0))
  (if (not (= i 10))
      (loop (+ i 1))))

(pp (table->list procedures-called))

;; output is the number of times each procedure was called:
;; ((#<procedure #6 table->list> . 1)
;;  (#<procedure #5 loop> . 10)
;;  (#<procedure #4 +> . 10)
;;  (#<procedure #3 not> . 11)
;;  (#<procedure #2 => . 11))

;;;----------------------------------------------------------------------------
```

Reply:
Using the stepper fail on the unit test `06-thread/mutex_race.scm`. There are too much redundant calls to `mutex-lock!` and `mutex-unlock!`.

This test takes too much time to complete with the call handler.

Reply:
Perhaps the number of iterations and/or the number of threads could be lowered so that the time is more reasonable.

Also, a mutex should probably be added to implement a critical section for the updating of the `procedures-called` table, otherwise the table may be incorrectly updated by the concurrent threads.

Reply:
I took note of you're last comment and included a mutex in the critical section for the updating of the `procedures-called` table.

* I updated the unit-tests so they take 1m30s when the procedure calls are recorded otherwise they run normally.
* I added a way to enable the recording of procedure calls, which is disable by default.

Here is the command to get the list of called procedures in a single unit-test.
```
$ gsi -e '(deifne ##enable-trace? #t)' unit-tests/03-number/asinh.scm
(%GAMBIT-COVERAGE% (+ . 12) (imag-part . 5) (for-each . 11) (- . 200) (sqrt . 6) (make-rectangular . 102) (* . 6) (with-exception-handler . 1) (asinh . 423) (type-exception? . 1) (conjugate . 200) (log . 6) (real-part . 5) (nan? . 9))